New Arcball controls

danielefornari opened the pull request:
**Description**

Added brand new ArcballControls inspired by the one of [Meshlab](https://www.meshlab.net).
This is part of my bachelor thesis project at [Visual Computing Lab](http://vcg.isti.cnr.it)

danielefornari committed:
Create ArcballControls.html

danielefornari committed:
Create misc_controls_interactivecontrolscomparison.html

danielefornari committed:
Updated with interactivecontrolscomparison

danielefornari committed:
Updated with ArcballControls doc

danielefornari committed:
Updated with interactivecontrolscomparison

danielefornari committed:
Create ArcballControls.html

danielefornari committed:
Create misc_controls_interactivecontrolscomparison.jpg

danielefornari committed:
Create ArcballControls.js

cignoni committed:
Updated documentation

danielefornari committed:
Added copy/paste state functions

danielefornari committed:
Added copy/paste functions in arcball menu

danielefornari committed:
Merge branch 'dev' of https://github.com/danielefornari/three.js into dev

danielefornari committed:
Fixed copy/paste state functions

danielefornari committed:
Fixed controls switch and updated arcball options gui

danielefornari committed:
Updated documentation

danielefornari committed:
Added copy and paste state functions

danielefornari committed:
Rename misc_controls_interactivecontrolscomparison.html to misc_controls_arcball.html

danielefornari committed:
Renamed interactivecontrolscomparison to arcball

danielefornari committed:
Update files.json

danielefornari committed:
Delete misc_controls_interactivecontrolscomparison.jpg

danielefornari committed:
Merge branch 'dev' of https://github.com/danielefornari/three.js into dev

danielefornari committed:
Create misc_controls_arcball.jpg

danielefornari committed:
Fixed Perspective to Orthographic camera transition

malomo commented:
The example, included into the pull request, is [here](https://danielefornari.github.io/three.js/examples/?q=arcball#misc_controls_arcball).

danielefornari committed:
Removed unused variables and minor optimizations

haywirez commented:
@danielefornari would liketo invite you to add this to https://github.com/pmndrs/three-stdlib! 💚

Mugen87 commented on examples/jsm/controls/ArcballControls.js:
```diff
@@ -0,0 +1,2495 @@
+		this._manager.on( 'triplepanmove', this.onTriplePanMove );
+		this._manager.on( 'triplepanend', this.onTriplePanEnd );
+
+		this.domElement.addEventListener( 'mousedown', this.onMouseDown );
```
Do you mind using pointer events instead of mouse events? We are in the middle of a migration to pointer events and it would be good to avoid adding code which depends on `mousedown`, `mousemove` etc..

Mugen87 commented on examples/jsm/controls/ArcballControls.js:
```diff
@@ -0,0 +1,2495 @@
+
+		if ( event.button == 1 && this.enabled && this.enablePan ) {
+	
+			event.preventDefault();
```
Please avoid the usage of `preventDefault()`. We also want to avoid these calls since they produce unexpected side effect for app level code.

danielefornari replied:
Do you have any alternatives to suggest? I would like to prevent window scrolling for operations which involves mouse wheel.

Mugen87 commented on examples/jsm/controls/ArcballControls.js:
```diff
@@ -0,0 +1,2495 @@
+	Matrix4,
+	MathUtils
+} from '../../../build/three.module.js';
+import {} from 'https://cdnjs.cloudflare.com/ajax/libs/hammer.js/2.0.8/hammer.js'
```
Is it possible to avoid the usage of `hammer.js`?

In any event, external libraries should not be included like that. If you really need it, you have to copy it into the `examples/jsm/libs` folder.

danielefornari replied:
I choosed `hammer.js` for recognizing multi-touch gestures for touch interaction and actually is my main recognizer for pointer events, that's why i would prefer to keep using it.

Mugen87 replied:
In other controls like `OrbitControls` we detect multi-touch gestures without a library. I have the feeling `hammer.js` could be a blocker for this PR.

mrdoob replied:
Indeed! Would be great if `ArcballControls` didn't have any dependencies 👍

As @Mugen87 said, you can use `OrbitControls` as reference:

https://github.com/mrdoob/three.js/blob/2f6358dd55956e4f3595f304b2b2c925551228d3/examples/jsm/controls/OrbitControls.js#L800-L874

malomo replied:
Hi, I am @danielefornari advisor on this project.
The problem here is having high level cross-browser support for touch gestures and mouse interaction and, as far as I know, right now, there is no clear way to do it easily without an external library. `hammer.js` is a temporary solution, and I would stick with it while we figure out what to do next.
In any case, for high level interaction, I think it makes a lot of sense to delegate the handling to an external library. Does anyone knows a better/more up-to-date lib for this purpose?

mrdoob replied:
> The problem here is having high level cross-browser support for touch gestures and mouse interaction and, as far as I know, right now, there is no clear way to do it easily without an external library.

Yes there is. [Pointer Events API](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events) unifies mouse events and touch events in a single API.

Mugen87 commented:
For testing: https://raw.githack.com/danielefornari/three.js/dev/examples/misc_controls_arcball.html

Mugen87 commented on examples/misc_controls_arcball.html:
```diff
@@ -0,0 +1,438 @@
+
+			import { ArcballControls } from './jsm/controls/ArcballControls.js';
+			import { OrbitControls } from './jsm/controls/OrbitControls.js';
+			import { TrackballControls } from './jsm/controls/TrackballControls.js';
```
In order to simplify the example, I suggest you remove both `OrbitControls` and `TrackballControls`.

At the homepage, the control examples just showcase a single control class for simplicity.

danielefornari committed:
Fixed issue with pasteState function

SNiTEBoBy reviewed (approved):
Merge pull requested

danielefornari committed:
Merge branch 'mrdoob:dev' into dev

danielefornari committed:
Replaced Hammer.js with PointerEvents

Mugen87 commented on examples/jsm/controls/ArcballControls.js:
```diff
@@ -0,0 +1,2611 @@
+		const EPS = 0.000001;
+
+		//check min/max parameters
+		if ( this.camera.type == 'OrthographicCamera' ) {
```
Please use this pattern:
```js
if ( this.camera.isOrthographicCamera ) {
```
You can detect perspective camera's via `isPerspectiveCamera`.

danielefornari committed:
Fixed double tap issue

danielefornari committed:
Fixed double tap issue

danielefornari committed:
Fixed an issue causing a singlepan detection when releasing two fingers from the screen

danielefornari committed:
Fixed page scrolling issue

danielefornari committed:
Merge branch 'mrdoob:dev' into dev

danielefornari committed:
Fixed an issue with rotation animation

danielefornari committed:
Merge branch 'mrdoob:dev' into dev

danielefornari commented:
`ArcballControls` has been updated to best fit the required changes: we replaced MouseEvent to PointerEvent API, removed any external dependencies and updated the example homepage.

danielefornari commented:
`ArcballControl` has been updated to best fit the required changes: we switched from MouseEvent to PointerEvent API, removed any external dependencies and updated the example homepage.

Mugen87 commented:
It seems you have to update the screenshot for the E2E test so the CI can pass.

danielefornari committed:
Updated example screenshot

Mugen87 commented on examples/misc_controls_arcball.html:
```diff
@@ -0,0 +1,268 @@
+								hdrEquirect.dispose();
+
+								//scene.background = hdrCubeRenderTarget.texture;
+								scene.environment = hdrCubeRenderTarget.texture;
```
No need to work with `PMREMGenerator` anymore. This component is now integrated in the renderer. Just do this:
```js
hdrEquirect.mapping = THREE.EquirectangularReflectionMapping;

scene.environment = hdrEquirect;

danielefornari committed:
removed PMREMGenerator component

cignoni commented:
So it is accepted and planned to be included the 133 release? That would be great! 😀

Mugen87 commented:
@mrdoob has still to review the PR but from my point of view it is good to go!

mrdoob commented:
Hmm, would it be possible to use the right mouse button for panning?
Right now is not possible to pan the model with a trackpad only (and no keyboard).

malomo commented:
> Hmm, would it be possible to use the right mouse button for panning?
> Right now is not possible to pan the model with a trackpad only (and no keyboard).

Right now panning is possible using the ctrl modifier (cmd on macOS)

mrdoob commented:
Yes, I know. But if you don't have a keyboard you can't pan.

danielefornari commented:
> Hmm, would it be possible to use the right mouse button for panning?
> Right now is not possible to pan the model with a trackpad only (and no keyboard).

We are thinking to overcome this limitation by adding a function which allows to associate a certain mouse/keyboard action with an operation. In this way one can also set mouse right button for panning.

mrdoob commented:
That sounds good to me 👍
Are you going to implement it in this PR?

danielefornari commented:
Yes, i will work on it these days

danielefornari committed:
Added functions to add/remove operation-mouse/key association and fixed an issue with touch inputs

danielefornari commented:
Functions to set/remove an association between an operation and a mouse/keyboard action has been added, so now inputs to perform operations can be customized anytime! Zoom/FOV manipulation can be also be associated with a mouse button, in this case the operation is performed with a vertical movement of the cursor.

danielefornari committed:
Merge branch 'mrdoob:dev' into dev

mrdoob commented:
Oh, I've just noticed that on ChromeOS I can pan with a trackpad using three fingers. I'll have to see if that works on MacOS.

Do you guys prefer three figers for panning instead of two fingers?

marcofugaro commented:
@mrdoob on MacOS by default three-finger pan is used to switch between fullscreen apps

https://user-images.githubusercontent.com/7217420/132357328-22ff4bc3-cb8e-443a-904d-61f1701a8373.mov

mrdoob commented:
Indeed...

@danielefornari can we set two finger / right mouse button for pan by default?

cignoni commented:
> Indeed...
> 
> @danielefornari can we set two finger / right mouse button for pan by default?

Not quite sure about these suggestions,
1) I still prefer to leave the right mouse button to context menu (something people always expects). Pan is currently mapped on middle mouse button (that is also an old tradition for the ones old enough to remember three button mouses :D) 
2) two finger movement on trackpad is often mapped to scroll, e.g. the same behaviour of the scroll wheel, so it more 'natural' to me to map two finger up/down drag to zoom (as it is in googlemaps or sketchfab for example)

mrdoob commented:
> 1) I still prefer to leave the right mouse button to context menu (something people always expects). Pan is currently mapped on middle mouse button (that is also an old tradition for the ones old enough to remember three button mouses :D) 

I know what you mean about leaving the right mouse button for the context menu, but that's a battle we went through 10 years ago already and we opted for losing the context menu in favour of better single-hand 3d scene control on MacOS.

> 2) two finger movement on trackpad is often mapped to scroll, e.g. the same behaviour of the scroll wheel, so it more 'natural' to me to map two finger up/down drag to zoom (as it is in googlemaps or sketchfab for example)

Sketchfab also uses right mouse button for panning:
https://sketchfab.com/3d-models/hand-painted-sketchfab-skater-girl-962a857810ff49b8a4a9b8ca71172384

danielefornari committed:
Modified default mouse actions

danielefornari commented:
As suggested, mouse right button has been setted for pan operation by default and, in order to best match Orbit/Trackball controls, wheel button has been setted for zoom operation

mrdoob commented:
Works beautifully now~

mrdoob commented:
Many thanks!
